### PR TITLE
Separate into 3 phases, Add Devsite Snippet Regions, Remove Install Script

### PR DESCRIPTION
**Pull request by waprin**
This PR is designed to allow Dev-site to include snippets from this tutorial:

https://cloud-dot-devsite.googleplex.com/appengine/articles/prediction_service_accounts

Since that tutorial is iterative, I created three directories:

phase1: initial skeleton
phase2: initial skeleton + sentiment analysis with hosted model
phase3: final project with language prediction via trained model

I also included the regions needed by this tutorial.

I removed the install script since we are now recommending using pip-t.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,14 +1,23 @@
 # prediction-appengine-guestbook-python
 
+## Phases
+This project is associated with this tutorial:
+
+https://cloud-dot-devsite.googleplex.com/appengine/articles/prediction_service_accounts
```
Comment (tmatsuo) [inline README.md]: This is the staging URL. Replace it with the real one.

File: README.md [hunk 2]
```diff
@@ -1,14 +1,23 @@
...
+
+Each directory represents a phase in the tutorial. phase1 is the basic sketon, phase2
```
Comment (tmatsuo) [inline README.md]: skeleton?

File: README.md [hunk 3]
```diff
@@ -1,14 +1,23 @@
...
 
 - Install Python-2.7 and virtualenv
 
 - Install google cloud SDK
 
-- Run scripts/setup.sh with the project directory as an argument.
+- Change into the directory of your choice and run
 
-  $ scripts/setup-google-api-client.sh .
+  $ pip install -r requirements.txt -t libs
```
Comment (tmatsuo) [inline README.md:20]: If the user goes through all 3 directories, they will end up downloading the libs 3 times. It's not a big deal, and also it's a general problem with our samples.

You can leave it for now...

File: phase1/app.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+api_version: 1
+threadsafe: true
+
+handlers:
+- url: /favicon\.ico
+  static_files: favicon.ico
+  upload: favicon\.ico
+
+- url: .*
+  script: main.APPLICATION
```
Comment (tmatsuo) [inline phase1/app.yaml]: I think there is a consensus about the variable name of the WSGI application.

EDIT:
I think we use 'app'.

Comment (theacodes) [inline phase1/app.yaml]: We do indeed use 'app' now.

File: phase1/main.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+import urllib
+
+from google.appengine.ext import ndb
+from google.appengine.api import users
+from google.appengine.ext.webapp import template
+import webapp2
+
+
+# Global variables
+API_KEY = "your-api-key"
```
Comment (tmatsuo) [inline phase1/main.py]: I think we should use the usable auth. You can use the App Engine service account on App Engine production, and you can use a generic service account on devserver (it only works with gcloud preview app IIRC)

File: phase2/appengine_config.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Prediction API sample application config."""
+
+import os
+import sys
+
+sys.path.append(os.path.join(os.path.dirname(__file__), 'libs'))
```
Comment (tmatsuo) [inline phase2/appengine_config.py]: You can use vendor.

``` python
from google.appengine.ext import vendor
```

or something

File: phase2/main.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+from apiclient.discovery import build
+import webapp2
+
+# Global variables
+API_KEY = "your-api-key"
+PROJECT_ID = "your-numeric-project-id"
+
+# [START credentials_setup]
+# Set up the Prediction API service
+CREDENTIALS = AppAssertionCredentials(
```
Comment (tmatsuo) [inline phase2/main.py]: ditto

File: phase2/main.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+SERVICES = threading.local()
+# [END credentials_setup]
+
+
+def get_service():
+    """Returns a prediction API service object local to the current thread."""
+    http = CREDENTIALS.authorize(httplib2.Http(memcache))
+    if not hasattr(SERVICES, "service"):
+        SERVICES.service = build("prediction", "v1.6", http=http,
+                                 developerKey=API_KEY)
```
Comment (tmatsuo) [inline phase2/main.py]: we may need to preserve API_KEY

File: phase2/appengine_config.py (hunk 6 above)
Comment (theacodes) [inline phase2/appengine_config.py]: Yes, please use vendor. https://cloud.google.com/appengine/docs/python/tools/libraries27#vendoring

File: phase3/app.yaml [hunk 9]
```diff
@@ -8,13 +8,13 @@ handlers:
...
   login: admin
 
 - url: /trainmodel
-  script: main.APPLICATION
-  login: admin  
-
-- url: .*
-  script: main.APPLICATION
+  script: main.app
+  login: admin
```
Comment (tmatsuo) [inline phase3/app.yaml]: add newline?

File: phase3/appengine_config.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Prediction API sample application config."""
+
+import os
+import sys
+
+sys.path.append(os.path.join(os.path.dirname(__file__), 'libs'))
```
Comment (tmatsuo) [inline phase3/appengine_config.py]: You can use vendor

File: phase3/main.py [hunk 11]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
 #
-# Copyright 2007 Google Inc.
+# Copyright 2015 Google Inc.
```
Comment (tmatsuo) [inline phase3/main.py:3]: Wow is it really that old?

**Comment by tmatsuo** [comment]
I just put the first batch of comments. Also make the travis build all green :)

File: phase2/main.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+from google.appengine.ext.webapp import template
+from oauth2client.appengine import AppAssertionCredentials
+from apiclient.discovery import build
+import webapp2
+
+PROJECT_ID = 0  # your numeric project id
+
+# [START credentials_setup]
+# Set up the Prediction API service
+CREDENTIALS = AppAssertionCredentials(
```
Comment (tmatsuo) [inline phase2/main.py]: You should use https://developers.google.com/accounts/docs/application-default-credentials
How about:

``` python
from oauth2client.client import GoogleCredentials

def get_credentials():
    credentials = GoogleCredentials.get_application_default()
    if credentials.create_scoped_required():
        credentials = credentials.create_scoped(BIGQUERY_SCOPES)

CREDENTIALS = get_credentials()
```

The code treats it as a constant so modifying the value seems weird to me, so how about to introduce a helper function?

File: tox.ini [hunk 13]
```diff
@@ -4,4 +4,4 @@ envlist = flake8-py2.7
 
 [testenv:flake8-py2.7]
 deps = flake8
-commands = flake8 main.py appengine_config.py
+commands = flake8 phase1/main.py phase2/main.py phase3/main.py
```
Comment (tmatsuo) [inline tox.ini]: I don't like what I did. I put all the files as flake8 arguments, but there should be a better way.
Can you use something like?

```
commands = flake8 --exclude=lib
```

File: phase3/app.yaml (hunk 9 above)
Comment (tmatsuo) [inline phase3/app.yaml]: I still see the no newline at the end of the file...

**Comment by tmatsuo** [comment]
I put the second batch! I hope this is the last iteration :)

File: README.md (hunk 3 above)
Comment (tmatsuo) [inline README.md:20]: One more thing, can you remove the API_KEY in "Edit main.py" section as well?

File: phase2/main.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+# [START credentials_setup]
+# Set up the Prediction API service
+credentials = GoogleCredentials.get_application_default()
+credentials = credentials.create_scoped(
+    'https://www.googleapis.com/auth/prediction'
+    ' https://www.googleapis.com/auth/devstorage.full_control')
+SERVICES = threading.local()
+# [END credentials_setup]
+
+# dont change this
```
Comment (tmatsuo) [inline phase2/main.py]: How about:

```
# Don't change this. This is the fixed project ID for Prediction API hosted mode.
```

File: phase3/main.py [hunk 15]
```diff
@@ -25,40 +25,49 @@
...
 # Global variables
-DATA_FILE = "your-bucket-name/language_id.txt"
+DATA_FILE = "your-bucket/language_id.txt"
 MODEL_ID = "your-model-id"  # it can be the same as the app_id
-API_KEY = "your-api-key"
-PROJECT_ID = "your-numeric-project-id"
+PROJECT_ID = 0  # put your numerical project-id here
+# [END global_variables]
+
+# dont change this
```
Comment (tmatsuo) [inline phase3/main.py]: ditto for the comment

File: phase2/main.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+import os
+import threading
+import urllib
+
+from google.appengine.ext import ndb
+from google.appengine.api import users
+from google.appengine.api import memcache
+from google.appengine.ext.webapp import template
+from oauth2client.client import GoogleCredentials
+from apiclient.discovery import build
```
Comment (tmatsuo) [inline phase2/main.py]: super nit:
Import order seems wrong (maybe alphabetical?).

File: phase3/main.py [hunk 17]
```diff
@@ -25,40 +25,49 @@
 from google.appengine.api import users
 from google.appengine.api import memcache
 from google.appengine.ext.webapp import template
-from oauth2client.appengine import AppAssertionCredentials
 from apiclient.discovery import build
 import webapp2
 
+from oauth2client.client import GoogleCredentials
```
Comment (tmatsuo) [inline phase3/main.py:29]: Import order?

**Comment by tmatsuo** [comment]
LGTM except for some nits, you can merge :)

File: phase2/main.py [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import httplib2
+import os
+import threading
+import urllib
+
+from apiclient.discovery import build
```
Comment (theacodes) [inline phase2/main.py]: Use googleapiclient instead of apiclient.

File: phase3/main.py [hunk 19]
```diff
@@ -21,44 +21,52 @@
 import threading
 import urllib
 
+from apiclient.discovery import build
```
Comment (theacodes) [inline phase3/main.py]: Same here. Use googleapiclient instead of apiclient.

**Comment by theacodes** [comment]
One nit from me as well. Don't let it block you though.

**Pull request merged**